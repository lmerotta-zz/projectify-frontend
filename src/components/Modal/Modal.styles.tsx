import tw, { styled } from "twin.macro";
import { motion } from "framer-motion";

export const Backdrop = motion(
  tw.div`fixed w-full h-full bg-primary p-4 flex justify-center items-center top-0 left-0 bottom-0 right-0 z-index[1050]`
);

export const Wrapper = motion(
  styled.div`
    ${tw`rounded-md bg-white shadow-2xl w-full h-auto relative z-10 md:w-4/5 lg:w-2/4 xl:w-2/5`}
  `
);
