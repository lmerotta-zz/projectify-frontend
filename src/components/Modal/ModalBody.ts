import tw from "twin.macro";

const ModalBody = tw.div`px-4 py-4`;

export default ModalBody;
