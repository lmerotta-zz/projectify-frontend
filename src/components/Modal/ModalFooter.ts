import tw, { styled } from "twin.macro";

const ModalFooter = styled.div<{ leftAlign?: boolean }>`
  ${tw`p-4 border-t-2 border-gray-200 flex`}
  ${({ leftAlign }) => (leftAlign ? "" : tw`justify-end`)}
`;

export default ModalFooter;
