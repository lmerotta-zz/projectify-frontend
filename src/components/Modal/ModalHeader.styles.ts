import { XIcon } from "@heroicons/react/outline";
import tw from "twin.macro";

export const ModalHeaderWrapper = tw.div`p-4 flex items-start justify-between border-b-2 border-gray-200`;
export const ModalHeaderTitle = tw.h5`text-lg font-thin text-dark flex-grow`;
export const ModalHeaderCloseButtonWrapper = tw.div`flex self-stretch flex-shrink-0 cursor-pointer relative top-1.5`;
export const ModalHeaderCloseButton = tw(XIcon)`w-4 h-4`;
