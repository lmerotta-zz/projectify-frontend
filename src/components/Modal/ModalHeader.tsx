/** @jsxImportSource @emotion/react */
import { ReactNode } from "react";
import * as Styles from "./ModalHeader.styles";

type ModalHeaderProps = {
  children?: ReactNode;
  onClose?: () => void;
};

const ModalHeader = ({ children, onClose }: ModalHeaderProps) => (
  <Styles.ModalHeaderWrapper>
    <Styles.ModalHeaderTitle>{children}</Styles.ModalHeaderTitle>
    {onClose && (
      <Styles.ModalHeaderCloseButtonWrapper
        data-testid="modal-close-button"
        onClick={() => onClose()}
      >
        <Styles.ModalHeaderCloseButton />
      </Styles.ModalHeaderCloseButtonWrapper>
    )}
  </Styles.ModalHeaderWrapper>
);

export default ModalHeader;
