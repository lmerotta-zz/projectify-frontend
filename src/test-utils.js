import React from "react";
import { render } from "@testing-library/react";
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";

const cache = createCache({ key: "iodajdoiwanoawdni" });
cache.compat = true;

const AllTheProviders = ({ children }) => {
  return <CacheProvider value={cache}>{children}</CacheProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
