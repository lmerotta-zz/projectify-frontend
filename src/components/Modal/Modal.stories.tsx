/** @jsxImportSource @emotion/react */
import { ComponentProps, useState } from "react";
import Modal from "./Modal";
import { Story } from "@storybook/react";
import ModalHeader from "./ModalHeader";
import ModalBody from "./ModalBody";
import ModalFooter from "./ModalFooter";
import { Button } from "components";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Modal",
  component: Modal,
  parameters: {
    docs: {
      inlineStories: false,
      iframeHeight: 500,
    },
  },
  argTypes: {
    isOpen: {
      control: true,
    },
    onClose: {
      control: false,
    },
  },
};

const Template: Story<ComponentProps<typeof Modal>> = (args) => {
  const [isOpen, setOpen] = useState(false);

  return (
    <div>
      <button onClick={() => setOpen(true)}>Open the modal</button>
      <Modal {...args} onClose={() => setOpen(false)} isOpen={isOpen}>
        <ModalHeader onClose={() => setOpen(false)}>
          Welcome to my modal!
        </ModalHeader>
        <ModalBody>
          <h5>You can put your content here</h5>
          <p>And any content!</p>
        </ModalBody>
        <ModalFooter>
          <Button onClick={() => setOpen(false)}>Submit</Button>
        </ModalFooter>
      </Modal>
    </div>
  );
};

export const Default = Template.bind({});
