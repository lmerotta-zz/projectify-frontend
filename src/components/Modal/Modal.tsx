/** @jsxImportSource @emotion/react */
import { ReactNode } from "react";
import { AnimatePresence } from "framer-motion";
import * as Styles from "./Modal.styles";

const backdropVariants = {
  open: {
    opacity: 1,
  },
  closed: {
    opacity: 0,
  },
};

const wrapperVariants = {
  open: {
    opacity: 1,
    y: 0,
  },
  closed: {
    opacity: 0,
    y: "-100%",
  },
};

type ModalProps = {
  onClose?: () => void;
  isOpen?: boolean;
  closeOnOutsideClick?: boolean;
  children?: ReactNode;
};

const Modal = ({
  onClose,
  isOpen,
  children,
  closeOnOutsideClick = false,
}: ModalProps) => (
  <AnimatePresence>
    {isOpen && (
      <Styles.Backdrop
        data-testid="modal-backdrop"
        key="modal-backdrop"
        variants={backdropVariants}
        initial="closed"
        animate="open"
        exit="closed"
        onClick={(e) => {
          /* istanbul ignore else */
          if (e.target === e.currentTarget) {
            closeOnOutsideClick && onClose?.();
          }
        }}
      >
        <Styles.Wrapper
          variants={wrapperVariants}
          initial="closed"
          animate="open"
          exit="closed"
        >
          {children}
        </Styles.Wrapper>
      </Styles.Backdrop>
    )}
  </AnimatePresence>
);

export default Modal;
