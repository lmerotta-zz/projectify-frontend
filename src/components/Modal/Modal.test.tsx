import { render, fireEvent } from "test-utils";
import Modal from "./Modal";
import ModalBody from "./ModalBody";
import ModalFooter from "./ModalFooter";

describe("Modal unit tests", () => {
  it("Closes on backdrop click", () => {
    const close = jest.fn();
    const result = render(
      <Modal onClose={close} closeOnOutsideClick isOpen>
        <ModalBody>This is the body</ModalBody>
        <ModalFooter>This is the footer</ModalFooter>
      </Modal>
    );

    fireEvent.click(result.getByTestId("modal-backdrop"));

    expect(close).toHaveBeenCalled();
    expect(result.asFragment()).toMatchSnapshot();
  });

  it("Does not close on backdrop click", () => {
    const close = jest.fn();
    const result = render(
      <Modal onClose={close} isOpen>
        <ModalBody>This is the body</ModalBody>
        <ModalFooter leftAlign>This is the footer</ModalFooter>
      </Modal>
    );

    fireEvent.click(result.getByTestId("modal-backdrop"));

    expect(close).not.toHaveBeenCalled();
    expect(result.asFragment()).toMatchSnapshot();
  });

  it("Does not render if not open", () => {
    const result = render(
      <Modal>
        <ModalBody>This is the body</ModalBody>
        <ModalFooter>This is the footer</ModalFooter>
      </Modal>
    );

    expect(() => result.getByTestId("modal-backdrop")).toThrow();
  });
});
