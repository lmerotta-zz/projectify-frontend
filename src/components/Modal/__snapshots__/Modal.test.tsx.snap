// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal unit tests Closes on backdrop click 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: fixed;
  width: 100%;
  height: 100%;
  --tw-bg-opacity: 1;
  background-color: rgba(147, 51, 234, var(--tw-bg-opacity));
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  top: 0px;
  left: 0px;
  bottom: 0px;
  right: 0px;
  z-index: 1050;
}

.emotion-1 {
  border-radius: 0.375rem;
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  --tw-shadow: 0 25px 50px -12px rgba(0, 0, 0, 0.25);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
  width: 100%;
  height: auto;
  position: relative;
  z-index: 10;
}

@media (min-width: 768px) {
  .emotion-1 {
    width: 80%;
  }
}

@media (min-width: 1024px) {
  .emotion-1 {
    width: 50%;
  }
}

@media (min-width: 1280px) {
  .emotion-1 {
    width: 40%;
  }
}

.emotion-2 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
}

.emotion-3 {
  padding: 1rem;
  border-top-width: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(229, 231, 235, var(--tw-border-opacity));
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

<div
    class="emotion-0"
    data-testid="modal-backdrop"
    style="opacity: 0;"
  >
    <div
      class="emotion-1"
      style="opacity: 0; transform: translateY(-100%) translateZ(0);"
    >
      <div
        class="emotion-2"
      >
        This is the body
      </div>
      <div
        class="emotion-3"
      >
        This is the footer
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal unit tests Does not close on backdrop click 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: fixed;
  width: 100%;
  height: 100%;
  --tw-bg-opacity: 1;
  background-color: rgba(147, 51, 234, var(--tw-bg-opacity));
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  top: 0px;
  left: 0px;
  bottom: 0px;
  right: 0px;
  z-index: 1050;
}

.emotion-1 {
  border-radius: 0.375rem;
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  --tw-shadow: 0 25px 50px -12px rgba(0, 0, 0, 0.25);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
  width: 100%;
  height: auto;
  position: relative;
  z-index: 10;
}

@media (min-width: 768px) {
  .emotion-1 {
    width: 80%;
  }
}

@media (min-width: 1024px) {
  .emotion-1 {
    width: 50%;
  }
}

@media (min-width: 1280px) {
  .emotion-1 {
    width: 40%;
  }
}

.emotion-2 {
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
}

.emotion-3 {
  padding: 1rem;
  border-top-width: 2px;
  --tw-border-opacity: 1;
  border-color: rgba(229, 231, 235, var(--tw-border-opacity));
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
    class="emotion-0"
    data-testid="modal-backdrop"
    style="opacity: 0;"
  >
    <div
      class="emotion-1"
      style="opacity: 0; transform: translateY(-100%) translateZ(0);"
    >
      <div
        class="emotion-2"
      >
        This is the body
      </div>
      <div
        class="emotion-3"
      >
        This is the footer
      </div>
    </div>
  </div>
</DocumentFragment>
`;
